src/jit/flowgraph.cpp

```diff
@@ -25718,3 +25720,333 @@ bool Compiler::fgUseThrowHelperBlocks()
+    int                          updateCount = 0;
+
+    for (; !iter.Equal(end); iter++)
+    {
+        BasicBlock* const nonCanonicalBlock = iter.Get();
+        BasicBlock* const canonicalBlock    = iter.GetValue();
+        flowList*         nextPredEdge      = nullptr;
+
+        // Walk pred list of the non canonical block, updating flow to target
+        // the canonical block instead.
+        for (flowList* predEdge = nonCanonicalBlock->bbPreds; predEdge != nullptr; predEdge = nextPredEdge)
+        {
+            BasicBlock* const predBlock = predEdge->flBlock;
+            nextPredEdge                = predEdge->flNext;
+
+            switch (predBlock->bbJumpKind)
+            {
+                case BBJ_NONE:
+                {
+                    fgTailMergeThrowsFallThroughHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_ALWAYS:
+                {
+                    fgTailMergeThrowsJumpToHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_COND:
+                {
+                    // Flow to non canonical block could be via fall through or jump or both.
+                    if (predBlock->bbNext == nonCanonicalBlock)
+                    {
+                        fgTailMergeThrowsFallThroughHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    }
+
+                    if (predBlock->bbJumpDest == nonCanonicalBlock)
+                    {
+                        fgTailMergeThrowsJumpToHelper(predBlock, nonCanonicalBlock, canonicalBlock, predEdge);
+                    }
+                    updateCount++;
+                }
+                break;
+
+                case BBJ_SWITCH:
+                {
+                    JITDUMP("*** " FMT_BB " now branching to " FMT_BB "\n", predBlock->bbNum, canonicalBlock->bbNum);
+                    fgReplaceSwitchJumpTarget(predBlock, canonicalBlock, nonCanonicalBlock);
+                    updateCount++;
+                }
+                break;
+
+                default:
+                    // We don't expect other kinds of preds, and it is safe to ignore them
+                    // as flow is still correct, just not as optimized as it could be.
+                    break;
+            }
+        }
+    }
+
+    // If we altered flow, reset fgModified. Given where we sit in the
+    // phase list, flow-dependent side data hasn't been built yet, so
+    // nothing needs invalidation.
+    //
+    // Note we could invoke a cleanup pass here, but optOptimizeFlow
+    // seems to be missing some safety checks and doesn't expect to
+    // see an already cleaned-up flow graph.
+    if (updateCount > 0)
+    {
+        assert(fgModified);
+        fgModified = false;
```

Comment:
This seems a little worrisome, but ok... Wouldn't (Shouldn't) the first code to generate "flow-dependent side data" reset this?

Reply:
Similar to logic in `fgCreateFuncletPrologBlocks`. 

The `fgModified` concept should probably be updated to just directly invalidate the dependent analyses, if any (in this case and the funclet case, there are none).